plugins/outputs/loki/stream.go

```diff
@@ -0,0 +1,71 @@
+package loki
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+)
+
+type (
+	Log []string
+
+	Streams map[string]*Stream
+
+	Stream struct {
+		Labels map[string]string `json:"stream"`
+		Logs   []Log             `json:"values"`
+	}
+
+	Request struct {
+		Streams []Stream `json:"streams"`
+	}
+)
+
+func (s Streams) insertLog(ts []*telegraf.Tag, l Log) {
+	key := uniqKeyFromTagList(ts)
+	_, ok := s[key]
+
+	if !ok {
+		s[key] = newStream(ts)
+	}
+
+	s[key].Logs = append(s[key].Logs, l)
+}
+
+func (s Streams) MarshalJSON() ([]byte, error) {
+	r := Request{
+		Streams: make([]Stream, 0, len(s)),
+	}
+
+	for _, stream := range s {
+		r.Streams = append(r.Streams, *stream)
+	}
```

Comment:
Just a heads-up: In golang access to maps is randomized, so if you rely on a certain order of streams you have to enforce this order after this loop. We had quite some problems with e.g. tests ignoring that fact and fail randomly.

As I said, no request for change bit just a heads-up.